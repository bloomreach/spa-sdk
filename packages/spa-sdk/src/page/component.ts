/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { MetaCollectionModel, Meta } from './meta';

/**
 * Generic component type.
 */
export const TYPE_COMPONENT = 'COMPONENT';

/**
 * Parameters of a component.
 */
export interface ComponentParameters {
  [name: string]: string | undefined;
}

/**
 * Meta-data of a component.
 */
export interface ComponentMeta extends MetaCollectionModel {
  params?: ComponentParameters;
}

interface Models {
  [model: string]: any;
}

interface Links {
  componentRendering?: { href: string };
}

/**
 * Model of a component.
 */
export interface ComponentModel {
  _meta?: ComponentMeta;
  _links?: Links;
  id: string;
  models?: Models;
  name?: string;
  type: typeof TYPE_COMPONENT | string;
  components?: ComponentModel[];
}

/**
 * A component in the current page.
 */
export interface Component {
  /**
   * @return The component meta-data collection.
   */
  getMeta(): Meta[];

  /**
   * @return The map of models.
   */
  getModels(): Models;

  /**
   * @return The link to the partial component model.
   */
  getModelUrl(): string | undefined;

  /**
   * @return The name of this component.
   */
  getName(): string;

  /**
   * @return the parameters of this component.
   */
  getParameters(): ComponentParameters;

  /**
   * Looks up for a nested component.
   * @param componentNames A lookup path.
   */
  getComponent<T extends Component>(): T;
  getComponent<T extends Component>(...componentNames: string[]): T | undefined;

  /**
   * Looks up for a nested component by its id.
   * @param id A component id.
   */
  getComponentById<T extends Component>(id: string): T | undefined;
}

export class Component implements Component {
  constructor(
    protected model: ComponentModel,
    protected children: Component[] = [],
    protected meta: Meta[] = [],
  ) {}

  protected getId() {
    return this.model.id;
  }

  getMeta() {
    return this.meta;
  }

  getModels() {
    return this.model.models || {};
  }

  getModelUrl() {
    return this.model._links && this.model._links.componentRendering && this.model._links.componentRendering.href;
  }

  getName() {
    return this.model.name || '';
  }

  getParameters() {
    return this.model._meta && this.model._meta.params || {};
  }

  getComponent<T extends Component>(): T;
  getComponent<T extends Component>(...componentNames: string[]): T | undefined;
  getComponent(...componentNames: string[]) {
    // tslint:disable-next-line:no-this-assignment
    let component: Component | undefined = this;

    while (componentNames.length && component) {
      const name = componentNames.shift()!;
      component = component.children.find(component => component.getName() === name);
    }

    return component;
  }

  getComponentById(id: string) {
    const queue = [this as Component];

    while (queue.length) {
      const component = queue.shift()!;

      if (component.model.id === id) {
        return component;
      }

      queue.push(...component.children);
    }
  }
}
