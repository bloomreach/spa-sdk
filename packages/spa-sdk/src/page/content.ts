/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Model of a content item.
 */
export interface ContentModel {
  id: string;
  localeString?: string;
  name: string;
  [property: string]: any;
}

/**
 * Content used on the page.
 */
export interface Content {
  /**
   * Returns the content id.
   */
  getId(): string;

  /**
   * Returns the content locale.
   */
  getLocale(): string | undefined;

  /**
   * Returns the content name.
   */
  getName(): string;

  /**
   * Returns the content data as it is returned in the Page Model API.
   */
  getData(): ContentModel;
}

export class Content {
  constructor(protected model: ContentModel) {}

  getId() {
    return this.model.id;
  }

  getLocale() {
    return this.model.localeString;
  }

  getName() {
    return this.model.name;
  }

  getData() {
    return this.model;
  }
}
