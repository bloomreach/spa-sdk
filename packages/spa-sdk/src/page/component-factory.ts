/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component, ComponentModel } from './component';

interface Task {
  model: ComponentModel;
  children?: Component[];
  siblings?: Component[];
}

type ComponentType<T extends ComponentModel> = T['type'];
type ComponentBuilder<T extends ComponentModel, U extends Component> = (model: T, children: U[]) => Component;

/**
 * A component factory producing components based on a type.
 */
export class ComponentFactory {
  private mapping = new Map<ComponentType<any>, ComponentBuilder<any, any>>();

  /**
   * Registers a component builder for the specified type.
   * @param type The component type.
   * @param builder The component builder.
   */
  register<T extends ComponentModel, U extends Component = Component>(
    type: ComponentType<T>,
    builder: ComponentBuilder<T, U>,
  ) {
    this.mapping.set(type, builder);

    return this;
  }

  /**
   * Produces a component based on the model
   * @param model The component model.
   */
  create(model: ComponentModel) {
    let component: Component;
    const queue = [{ model } as Task];

    while (queue.length) {
      const head = queue.shift()!;
      if (!head.children && head.model.components && head.model.components.length) {
        head.children = [];
        queue.unshift(
          ...head.model.components.map(model => ({ model, siblings: head.children })),
          head,
        );

        continue;
      }

      component = this.buildComponent(head.model, head.children || []);

      if (head.siblings) {
        head.siblings.push(component);
      }
    }

    return component!;
  }

  private buildComponent(model: ComponentModel, children: Component[]) {
    const builder = this.mapping.get(model.type);
    if (!builder) {
      throw new Error(`Unsupported component type: '${model.type}'.`);
    }

    return builder(model, children);
  }
}
