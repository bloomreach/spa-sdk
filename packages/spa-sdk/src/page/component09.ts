/*
 * Copyright 2020 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { inject, injectable } from 'inversify';
import { ComponentChildrenToken, ComponentModelToken, ComponentMeta, Component } from './component';
import { Link } from './link';
import { MetaCollectionFactory } from './meta-collection-factory';
import { MetaCollection } from './meta-collection';
import { UrlBuilderService, UrlBuilder } from '../url';

/**
 * Generic component type.
 */
export const TYPE_COMPONENT = 'COMPONENT';

/**
 * Container item type.
 */
export const TYPE_COMPONENT_CONTAINER_ITEM = 'CONTAINER_ITEM_COMPONENT';

/**
 * Container type.
 */
export const TYPE_COMPONENT_CONTAINER = 'CONTAINER_COMPONENT';

export type ComponentType =
  | typeof TYPE_COMPONENT
  | typeof TYPE_COMPONENT_CONTAINER_ITEM
  | typeof TYPE_COMPONENT_CONTAINER;

type ComponentLinks = 'componentRendering';

type ComponentModels = Record<string, any>;

/**
 * Model of a component.
 */
export interface ComponentModel {
  _links: Record<ComponentLinks, Link>;
  _meta: ComponentMeta;
  id: string;
  models?: ComponentModels;
  name?: string;
  type: ComponentType;
  components?: ComponentModel[];
}

@injectable()
export class ComponentImpl implements Component {
  protected meta: MetaCollection;

  constructor(
    @inject(ComponentModelToken) protected model: ComponentModel,
    @inject(ComponentChildrenToken) protected children: Component[],
    @inject(MetaCollectionFactory) metaFactory: MetaCollectionFactory,
    @inject(UrlBuilderService) private urlBuilder: UrlBuilder,
  ) {
    this.meta = metaFactory(this.model._meta);
  }

  getId(): string {
    return this.model.id;
  }

  getMeta(): MetaCollection {
    return this.meta;
  }

  getModels<T extends ComponentModels>(): T;

  getModels(): Record<string, unknown> {
    return this.model.models || {};
  }

  getUrl(): string {
    return this.urlBuilder.getApiUrl(this.model._links.componentRendering.href!);
  }

  getName(): string {
    return this.model.name || '';
  }

  getParameters<T>(): T {
    return (this.model._meta.params ?? {}) as T;
  }

  getChildren(): Component[] {
    return this.children;
  }

  getComponent(): this;

  getComponent<U extends Component>(...componentNames: string[]): U | undefined;

  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
  getComponent(...componentNames: string[]) {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    let component: Component | undefined = this;

    while (componentNames.length && component) {
      const name = componentNames.shift()!;
      component = component.getChildren().find((childComponent) => childComponent.getName() === name);
    }

    return component;
  }

  getComponentById<U extends Component>(id: string): U | this | undefined;

  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
  getComponentById(id: string) {
    const queue = [this as Component];

    while (queue.length) {
      const component = queue.shift()!;

      if (component.getId() === id) {
        return component;
      }

      queue.push(...component.getChildren());
    }

    return undefined;
  }
}

/**
 * Checks whether a value is a page component.
 * @param value The value to check.
 */
export function isComponent(value: any): value is Component {
  return value instanceof ComponentImpl;
}
