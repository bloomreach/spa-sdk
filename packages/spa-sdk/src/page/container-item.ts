/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Typed } from 'emittery';
import { ComponentImpl, ComponentMeta, ComponentModel, ComponentParameters, Component } from './component';
import { EmitterMixin, Emitter } from '../emitter';
import { Events, PageUpdateEvent } from '../events';
import { Meta } from './meta';

/**
 * Container item type.
 */
export const TYPE_COMPONENT_CONTAINER_ITEM = 'CONTAINER_ITEM_COMPONENT';
const PARAMETER_HIDDEN = 'com.onehippo.cms7.targeting.TargetingParameterUtil.hide';

/**
 * Parameters of a container item.
 */
export interface ContainerItemParameters extends ComponentParameters {
  [PARAMETER_HIDDEN]?: 'on' | 'off';
}

/**
 * Meta-data of a container item.
 */
export interface ContainerItemMeta extends ComponentMeta {
  params?: ContainerItemParameters;
  paramsInfo?: ComponentParameters;
}

/**
 * Model of a container item.
 */
export interface ContainerItemModel extends ComponentModel {
  _meta?: ContainerItemMeta;
  label?: string;
  type: typeof TYPE_COMPONENT_CONTAINER_ITEM;
}

/**
 * Container item update event.
 */
export interface ContainerItemUpdateEvent {}

interface ContainerItemEvents {
  update: ContainerItemUpdateEvent;
}

/**
 * A component that can be configured in the UI.
 */
export interface ContainerItem extends Component, Emitter<ContainerItemEvents> {
  /**
   * Returns the type of a container item. The available types depend on which
   * container items have been configured in the backend.
   *
   * @return The type of a container item (e.g. "Banner").
   */
  getType(): string | undefined;

  /**
   * Returns whether the component should not render anything.
   * Hiding components is only possible with the Relevance feature.
   *
   * @return Whether the component is hidden or not.
   */
  isHidden(): boolean;
}

export class ContainerItemImpl
  extends EmitterMixin<typeof ComponentImpl, ContainerItemEvents>(ComponentImpl)
  implements ContainerItem
{
  constructor(protected model: ContainerItemModel, eventBus: Typed<Events>, meta: Meta[] = []) {
    super(model, [], meta);
    eventBus.on('page.update', this.onPageUpdate.bind(this));
  }

  protected onPageUpdate(event: PageUpdateEvent) {
    const component = event.page.getComponent();
    if (!(component instanceof ContainerItemImpl) || component === this || component.getId() !== this.getId()) {
      return;
    }

    this.meta = component.meta;
    this.model = component.model;
    this.emit('update', {});
  }

  getType() {
    return this.model.label;
  }

  isHidden() {
    return !!(this.model._meta && this.model._meta.params && this.model._meta.params[PARAMETER_HIDDEN] === 'on');
  }

  getParameters(): ContainerItemParameters {
    return this.model._meta && this.model._meta.paramsInfo || {};
  }
}

/**
 * Checks whether a value is a container item.
 * @param value The value to check.
 */
export function isContainerItem(value: any): value is ContainerItem {
  return value instanceof ContainerItemImpl;
}
