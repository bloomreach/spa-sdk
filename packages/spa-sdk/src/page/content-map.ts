/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Content } from './content';

/**
 * The map holding content objects.
 */
export interface ContentMap extends Map<string, Content> {
  /**
   * Gets a content item by a reference.
   * @param reference The reference to a content item.
   */
  get(reference: string): Content | undefined;

  /**
   * Puts a content item under a reference.
   * @param reference The reference to a content item.
   * @param content The content.
   */
  set(reference: string, content: Content): this;
}
