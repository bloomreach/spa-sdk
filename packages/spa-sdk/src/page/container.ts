/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component, ComponentModel } from './component';
import { ContainerItem } from './container-item';
import { Meta } from './meta';

/**
 * Container type.
 */
export const TYPE_COMPONENT_CONTAINER = 'CONTAINER_COMPONENT';

export const TYPE_CONTAINER_BOX = 'hst.vbox';
export const TYPE_CONTAINER_UNORDERED_LIST = 'hst.unorderedlist';
export const TYPE_CONTAINER_ORDERED_LIST = 'hst.orderedlist';
export const TYPE_CONTAINER_INLINE = 'hst.span';
export const TYPE_CONTAINER_NO_MARKUP = 'hst.nomarkup';

type ContainerType = typeof TYPE_CONTAINER_BOX
  | typeof TYPE_CONTAINER_UNORDERED_LIST
  | typeof TYPE_CONTAINER_ORDERED_LIST
  | typeof TYPE_CONTAINER_INLINE
  | typeof TYPE_CONTAINER_NO_MARKUP;

/**
 * Model of a container item.
 */
export interface ContainerModel extends ComponentModel {
  type: typeof TYPE_COMPONENT_CONTAINER;
  xtype: ContainerType;
}

/**
 * A component that holds an ordered list of container item components.
 */
export interface Container extends Component {
  /**
   * Returns the type of a container.
   *
   * @see https://documentation.bloomreach.com/library/concepts/template-composer/channel-editor-containers.html
   * @return The type of a container (e.g. `TYPE_CONTAINER_BOX`).
   */
  getType(): ContainerType;

  /**
   * @return The children of a container.
   */
  getChildren(): ContainerItem[];
}

export class Container extends Component implements Container {
  constructor(
    protected model: ContainerModel,
    protected children: ContainerItem[] = [],
    protected meta: Meta[] = [],
  ) {
    super(model, children, meta);
  }

  getChildren() {
    return this.children;
  }

  getType() {
    return this.model.xtype;
  }
}
