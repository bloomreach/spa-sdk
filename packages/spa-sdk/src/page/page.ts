/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Typed } from 'emittery';
import { Component, ComponentMeta, ComponentModel } from './component';
import { ContentModel, Content } from './content';
import { ContentMap } from './content-map';
import { Events, PageUpdateEvent } from '../events';
import { Reference, isReference } from './reference';

/**
 * Meta-data of a page root component.
 */
export interface RootMeta extends ComponentMeta {
  pageTitle?: string;
}

/**
 * Model of a page root component.
 */
export interface RootModel extends ComponentModel {
  _meta?: RootMeta;
}

/**
 * Meta-data of a page.
 */
export interface PageMeta {
  preview?: boolean;
}

/**
 * Model of a page.
 */
export interface PageModel {
  _meta?: PageMeta;
  content?: { [reference: string]: ContentModel };
  page: RootModel;
}

/**
 * The current page to render.
 */
export interface Page {
  /**
   * Gets a component in the page (e.g. getComponent('main', 'right')).
   * Without any arguments it returns the root component.
   *
   * @param componentNames the names of the component and its parents.
   * @return The component, or `undefined` if no such component exists.
   */
  getComponent<T extends Component>(): T;
  getComponent<T extends Component>(...componentNames: string[]): T | undefined;

  /**
   * Gets a content item used in the page.
   * @param reference The reference to the content. It can be an object containing
   * an [RFC-6901](https://tools.ietf.org/html/rfc6901) JSON Pointer.
   */
  getContent(reference: Reference | string): Content;

  /**
   * Returns the title of the page, if configured.
   * @returns the title of the page, or `undefined` if not configured.
   */
  getTitle(): string | undefined;
}

export class Page implements Page {
  constructor(
    protected model: PageModel,
    protected root: Component,
    protected content: ContentMap,
    eventBus: Typed<Events>,
  ) {
    eventBus.on('page.update', this.onPageUpdate.bind(this));
  }

  protected onPageUpdate(event: PageUpdateEvent) {
    event.page.content.forEach((content, reference) => this.content.set(reference, content));
  }

  private static getContentReference(reference: Reference) {
    return  reference.$ref.split('/', 3)[2] || '';
  }

  getComponent<T extends Component>(): T;
  getComponent<T extends Component>(...componentNames: string[]): T | undefined;
  getComponent(...componentNames: string[]) {
    return this.root.getComponent(...componentNames);
  }

  getContent(reference: Reference | string) {
    const contentReference = isReference(reference)
      ? Page.getContentReference(reference)
      : reference;

    return this.content.get(contentReference);
  }

  getTitle() {
    return this.model.page._meta && this.model.page._meta.pageTitle;
  }
}
