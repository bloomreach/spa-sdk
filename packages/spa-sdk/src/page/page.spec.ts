/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Typed } from 'emittery';
import { ComponentImpl, Component, TYPE_COMPONENT } from './component';
import { ContentMap } from './content-map';
import { ContentImpl } from './content';
import { Events } from '../events';
import { PageImpl, Page } from './page';

describe('PageImpl', () => {
  let content: ContentMap;
  let eventBus: Typed<Events>;
  let root: Component;

  beforeEach(() => {
    content = new Map();
    eventBus = new Typed<Events>();
    root = new ComponentImpl({ id: 'id', type: TYPE_COMPONENT });

    jest.spyOn(root, 'getComponent');
    jest.spyOn(content, 'get');
  });

  describe('getComponent', () => {
    it('should forward a call to the root component', () => {
      const page = new PageImpl({ page: { id: 'id', type: TYPE_COMPONENT } }, root, content, eventBus);
      page.getComponent('a', 'b');

      expect(root.getComponent).toBeCalledWith('a', 'b');
    });
  });

  describe('getContent', () => {
    let page: Page;

    beforeEach(() => {
      page = new PageImpl({ page: { id: 'id', type: TYPE_COMPONENT } }, root, content, eventBus);
    });

    it('should resolve a reference', () => {
      page.getContent({ $ref: '/content/some-content' });

      expect(content.get).toBeCalledWith('some-content');
    });

    it('should pass a string reference directly', () => {
      page.getContent('content-reference');

      expect(content.get).toBeCalledWith('content-reference');
    });

    it('should return a content item', () => {
      const someContent = new ContentImpl({ id: 'some-id', name: 'some-name' });
      content.set('some-content', someContent);

      expect(page.getContent('some-content')).toBe(someContent);
    });
  });

  describe('getTitle', () => {
    it('should return a page title', () => {
      const page = new PageImpl(
        {
          page: {
            id: 'id',
            type: TYPE_COMPONENT,
            _meta: { pageTitle: 'something' },
          },
        },
        root,
        content,
        eventBus,
      );

      expect(page.getTitle()).toBe('something');
    });

    it('should return an undefined value', () => {
      const page1 = new PageImpl({ page: { id: 'id', type: TYPE_COMPONENT, _meta: {} } }, root, content, eventBus);
      const page2 = new PageImpl({ page: { id: 'id', type: TYPE_COMPONENT } }, root, content, eventBus);

      expect(page1.getTitle()).toBeUndefined();
      expect(page2.getTitle()).toBeUndefined();
    });
  });

  describe('onPageUpdate', () => {
    it('should update content on page.update event', async () => {
      const model = { page: { id: 'id', type: TYPE_COMPONENT } };
      const page = new PageImpl(model, root, content, eventBus);
      const someContent = new ContentImpl({ id: 'some-id', name: 'some-name' });

      expect(page.getContent('some-content')).toBeUndefined();
      await eventBus.emitSerial('page.update', {
        page: new PageImpl(model, root, new Map([['some-content', someContent]]), eventBus),
      });
      expect(page.getContent('some-content')).toBe(someContent);
    });
  });
});
