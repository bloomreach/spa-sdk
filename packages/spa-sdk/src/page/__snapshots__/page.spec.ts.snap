// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageImpl sanitize should keep rel, data-type, title, target, name and href attributes in anchor 1`] = `
"
        <div>
          <h1>Hello, World!</h1>
          <a rel=\\"noopener\\" data-type=\\"internal\\" title=\\"foo\\" target=\\"_blank\\" href=\\"https://www.example.com/\\" name=\\"use\\">use</a>
          <div></div>
        </div>"
`;

exports[`PageImpl sanitize should sanitize html 1`] = `
"
        <div>
          <h1>Hello, World!</h1>
          <p>Sanitize before <a href=\\"https://www.example.com/\\" name=\\"use\\">use</a></p>
          <div></div>
        </div>"
`;
